torch/_torch_docs.py

```diff
@@ -4924,6 +4950,8 @@ def merge_dicts(*dicts):
 .. note:: This function does not :ref:`broadcast <broadcasting-semantics>`.
           For broadcasting matrix products, see :func:`torch.matmul`.
 
+This function supports backward for strided matrices only.
```

Comment:
This brings up a few questions:

- If torch.mm supports sparse tensors, how should we say that? Should it be another property in our new "Supports..." statements, like the one [torch.sub](https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub) has? Then this might "Supports ..., sparse tensors and sparse autograd, ..."
- Are we being consistent about whether we use the term "strided" or "dense" to describe the "default" tensors?
- Do we want people to use torch.mm on sparse tensors or torch.sparse.smm?

Reply:
> This brings up a few questions:
> 
> * If torch.mm supports sparse tensors, how should we say that? Should it be another property in our new "Supports..." statements, like the one [torch.sub](https://pytorch.org/docs/master/generated/torch.sub.html?highlight=sub#torch.sub) has? Then this might "Supports ..., sparse tensors and sparse autograd, ..."

`torch.mm` does not support sparse autograd. So, the statement would read:

"Supports strided and sparse 2-D tensors as inputs, autograd with respect to strided inputs."

> * Are we being consistent about whether we use the term "strided" or "dense" to describe the "default" tensors?

This is a fair question and the answer is, in general, no: we have `to_dense` method and `strided` layout, for instance.

The word "strided" is used to describe the storage format while "dense" and "sparse" are adjectives that ought to describe the structure of data (characterized by some ratio of specified/unspecified tensor elements). While formally, strided tensors are always dense, then one could consider a sparse tensor being dense when the number of specified elements, `nse`, is large compared to `product(shape)`. However, it appears that people consider "dense tensors" to be always "strided tensors", although it is inaccurate, imho.

Notice that currently, we have one sparse tensor format with ``sparse_coo`` but soon we likely have more (as ``sparse_gcs`` from https://github.com/pytorch/pytorch/pull/44190), and each sparse format may implement a different set of supported properties.

In docs, I would suggest using the storage layout name rather than dense or sparse adjectives because soon the usage of dense/sparse will not be sufficient for distinguishing the different tensor implementations.

> * Do we want people to use torch.mm on sparse tensors or torch.sparse.smm?

Neither. People ought to choose the proper tools themselves based on their data and the desired output.
The table in "Linear Algebra operations" ought to help to make this decision.

Reply:
Something like that support statement seems reasonable. 

Being more vigilant about using "strided" instead of "dense" sounds good. Seems like maybe the to_dense function should be deprecated in favor of a to_strided?

I appreciate that we will soon have another sparse format, which is cool and this is trying to lay the groundwork for that change, but it may be simpler to focus on the one format we do have for the moment. 

As for picking the right tool, see my comment about that list of linear algebra operations. I worry about how intuitive each of those five options for performing a matmul are, and adding a couple aliases doesn't make it clearer.

Reply:
> Something like that support statement seems reasonable.
fixed.
 
> Being more vigilant about using "strided" instead of "dense" sounds good. Seems like maybe the to_dense function should be deprecated in favor of a to_strided?

I would take this suggestion even further: we currently have `torch.Tensor.to` method for dtype and/or device conversion and I would suggest overloading the `to` method for the layout conversion as well.
For instance, `to_dense()` would read `to(torch.strided)` and `to_sparse` would read `to(torch.sparse_coo)`, etc.

> I appreciate that we will soon have another sparse format, which is cool and this is trying to lay the groundwork for that change, but it may be simpler to focus on the one format we do have for the moment.

Notice that the groundwork for introducing new layouts is already built into pytorch internals. Only the docs are lagging behind of hinting that such extensions are possible. So, I slightly disagree with the suggested approach but I can understand that in pytorch it makes sense to strictly reflect only the currently available feature set. 

> As for picking the right tool, see my comment about that list of linear algebra operations. I worry about how intuitive each of those five options for performing a matmul are, and adding a couple aliases doesn't make it clearer.

I agree. https://github.com/pytorch/pytorch/issues/45897 aims at improving this situation.

Reply:
The rewritten sparse.rst now concentrates more on the existing COO sparse tensor support.
So, resolving this discussion.